src/html5_parser/element_class.rs

```diff
@@ -0,0 +1,157 @@
+use std::collections::HashMap;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct ElementClass {
+    /// a map of classes applied to an HTML element.
+    /// key = name, value = is_active
+    /// the is_active is used to toggle a class (JavaScript API)
+    class_map: HashMap<String, bool>,
+}
+
+impl Default for ElementClass {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ElementClass {
+    /// Initialise a new (empty) ElementClass
+    pub fn new() -> Self {
+        ElementClass {
+            class_map: HashMap::new(),
+        }
+    }
+
+    /// Initialize a class from a class string
+    /// with space-delimited class names
+    pub fn from_string(class_string: &str) -> Self {
+        let mut class_map_local = HashMap::new();
+        let classes = class_string.split_whitespace();
+        for class_name in classes {
+            class_map_local.insert(class_name.to_owned(), true);
+        }
+
+        ElementClass {
+            class_map: class_map_local,
+        }
+    }
+
+    /// Count the number of classes (active or inactive)
+    /// assigned to an element
+    pub fn len(&self) -> usize {
+        self.class_map.len()
+    }
+
+    /// Check if any classes are present
+    pub fn is_empty(&self) -> bool {
+        self.class_map.is_empty()
+    }
+
+    /// Check if class name exists
+    pub fn contains(&self, name: &str) -> bool {
+        self.class_map.contains_key(name)
+    }
+
+    /// Add a new class (if already exists, does nothing)
+    pub fn add(&mut self, name: &str) {
+        // by default, adding a new class will be active.
+        // however, map.insert will update a key if it exists
+        // and we don't want to overwrite an inactive class to make it active unintentionally
+        // so we ignore this operation if the class already exists
+        if !self.contains(name) {
```

Comment:
Probably want to keep an eye on this kind of substring matching.  Might be fine indefinitely, or might cause problems later on.